src/frontend/query_commands.ml

```diff
@@ -275,6 +275,19 @@ let dispatch pipeline (type a) : a Query_protocol.t -> a =
     to_string ()
 
   | Stack_or_heap_enclosing (pos, index) ->
+    (* Optimise allocations only on programs that have type-checked. *)
+    let errors =
+      Mpipeline.reader_lexer_errors pipeline @
+      Mpipeline.reader_parser_errors pipeline @
+      Mpipeline.ppx_errors pipeline @
+      Mpipeline.typer_errors pipeline
+    in
+    if not @@ List.exists errors
+        ~f:(function
+            | Msupport.Warning _ -> false
+            | _ -> true)
+    then Typecore.optimise_allocations ();
```

Comment:
I think (but am not sure) that this is unnecessary. I see `Mtyper.run` that calls `type_implementation`. Every route through `type_implementation` calls `optimise_allocations`. (Unless, of course, an exception is thrown!) So I think this call here is redundant.

Reply:
I had thought the same thing, but I was getting unexpected results so I did some print statement debugging. If you add something like `print_endline "optimising"` to `Typecore.optimise_allocations` and comment out these lines, you'll see that it never runs. And if you comment the lines back in, it'll only run once. (You can easily verify this by running `dune build && dune exec ocamlmerlin -- single stack-or-heap-enclosing -position 5:9 -filename tests/test-dirs/stack-or-heap.t/constructors.ml -ocamllib-path /j/office/app/ocaml/prod/5.1.1minus/5.1.1minus-13-d49e4649/el8+fl2/lib/ocaml/ < tests/test-dirs/stack-or-heap.t/constructors.ml`)

Reply:
I'd like to understand this better -- seems quite strange. At a minimum, we'll learn more about the flow of control through merlin. Could you maybe add some debugging to `type_implementation` and see where things stop progressing?

Reply:
(To be clear, more than ~2 hours spent on this query seems like a poor use of time. But if both of us are surprised here, let's learn a little more.)

Reply:
I think there was some confusion here because there are two functions called `type_implementation`. `Mtyper.type_implementation` (in `src/kernel/mtyper.ml`) is the one that is called by `Mtyper.run`. There is also `Typemod.type_implementation` (in `src/ocaml/typing/typemod.ml`) which calls `optimise_allocations` in every path but is never called, explaining why `optimise_allocations` is never called.

Reply:
Aha. Yes, this makes sense. Might it make sense also to add `optimise_allocations` to `Mtyper.type_implementation`? Perhaps right after the call to `type_structure`. We should check that this doesn't break anything, but it would seem to be the right place to do it, and doing it there would simplify the implementation over the current error check.

Reply:
Making this change results in `optimise_allocations` being called for other Merlin analyses besides `stack-or-heap-enclosing`, which causes some odd behavior.  For example, in the below program:
```
module X = struct
  type t = { z : int }
end

let f1 _ (x : X.t) = x.z

let f2 (z : string) (x : X.t) = x.X.z

let f3 (z : string) (x : X.t) = x.z
```
`type-enclosing` reports that `f3` has type `string -> local_ once_ (X.t -> int)`. (without the `optimise_allocations` call, it gives `string -> X.t -> int`)

Reply:
OK. Let sleeping dogs lie.